### PR TITLE
[mlir] Fix loop pipelining when the operand of `yield` is not defined in the loop body

### PR DESCRIPTION
**Pull request by Jokeren**

### CODE REVIEW

**Comment by Jokeren** [comment]
@ThomasRaoux

File: mlir/lib/Dialect/SCF/Transforms/LoopPipelining.cpp [hunk 1]
```diff
@@ -685,14 +696,15 @@ FailureOr<ForOp> mlir::scf::pipelineForLoop(RewriterBase &rewriter, ForOp forOp,
...
-    returnValues = pipeliner.emitEpilogue(rewriter);
+    pipeliner.emitEpilogue(rewriter, returnValues);
   }
   // 5. Erase the original loop and replace the uses with the epilogue output.
   if (forOp->getNumResults() > 0)
     rewriter.replaceOp(forOp, returnValues);
   else
     rewriter.eraseOp(forOp);
 
+  llvm::errs() << *newForOp->getParentOp() << "\n";
```
Comment (ThomasRaoux) [inline mlir/lib/Dialect/SCF/Transforms/LoopPipelining.cpp:707]: left over debug?

File: mlir/lib/Dialect/SCF/Transforms/LoopPipelining.cpp [hunk 2]
```diff
@@ -551,21 +560,24 @@ LogicalResult LoopPipelinerInternal::createKernel(
...
+    assert(def && "Only support loop carried dependencies of distance of 1 or "
+                  "defined outside the loop");
+    auto defStage = stages.find(def);
+    if (defStage != stages.end() && defStage->second > 0)
       setValueMapping(forOp.getRegionIterArgs()[retVal.index()],
                       newForOp->getResult(retVal.index()),
-                      maxStage - defStage + 1);
-    }
+                      maxStage - defStage->second + 1);
+    else
```
Comment (ThomasRaoux) [inline mlir/lib/Dialect/SCF/Transforms/LoopPipelining.cpp:570]: nit: please add  braces for multi-line if/else

**Review by ThomasRaoux — COMMENTED** [review]
Looks good to me overall, added few minor comments

**Comment by Jokeren** [comment]
Thank you Thomas!

**Review by ThomasRaoux — APPROVED** [review]
LGTM

**Pull request merged**